### PR TITLE
Add PreorderReasoning comforts

### PR DESCRIPTION
**Pull request by crisoagf**
PR for #2776

### CODE REVIEW

**Comment by crisoagf** [comment]
FWIW, I compiled all libs with this change and everything seems to work.

**Comment by ohad** [comment]
Do we really need to use type-classes to do this?
A principle I learned for @gallais: don't search when you can compute.
We know that we need a transitive relation for 2-steps or more, and we know we need a reflexive relation for 0-steps, and we know we don't need anything for 1 step: we already have the proof.

So there's no need to search for the CalcType instance (which might produce ambiguities): we can compute it directly.

Here's the code:
```idris
-- Using a McKinna-McBride view records the depenencies on the type of derivation
-- the additional benefits for pattern-matching aren't used here
public export
data DerivationType : FastDerivation leq x y -> Type where
  TrivialDerivation : DerivationType (|~ x)
  SingleStepDerivation : DerivationType (|~ x <~ step)
  NonTrivialDerivation : DerivationType (der <~ step1 <~ step2)

public export
derivationType : (der : FastDerivation leq x y) -> DerivationType der
derivationType (|~ x) = TrivialDerivation
derivationType (|~ x <~ step) = SingleStepDerivation
derivationType ((z <~ w) <~ step) = NonTrivialDerivation

public export
0 Prerequisite : {0 der : FastDerivation {a = dom} leq x y} -> DerivationType der -> Type
Prerequisite TrivialDerivation = Reflexive dom leq
Prerequisite SingleStepDerivation = ()
Prerequisite NonTrivialDerivation = Transitive dom leq

-- Once we have the prerequisite for transitivity, we have the
-- prerequisite for its inductive predecessor:
public export
inductivePrerequisite : (der : FastDerivation leq x y) ->
  (0 step1 : Step leq y z) -> (0 step2 : Step leq z w) ->
  Prerequisite (derivationType (der <~ step1 <~ step2)) ->
  Prerequisite (derivationType (der <~ step1))
inductivePrerequisite (  |~ y    ) step1 step2 prerequisite = ()
inductivePrerequisite (z <~ step0) step1 step2 prerequisite = prerequisite

public export
preorderPrerequisite : Preorder dom leq => (der : FastDerivation leq x y) -> Prerequisite (derivationType der)
preorderPrerequisite (|~ x) = %search
preorderPrerequisite (|~ x <~ step) = ()
preorderPrerequisite (der <~ step1 <~ step2) = %search

||| The Prerequisite for the derivation:
||| 0-length derivation: Reflexive dom leq
||| 1-length derivation: Unit (no prerequisite)
||| 2 steps of longer: Transitivity
public export
CalcSmart : {0 x : dom} -> {0 y : dom} -> (der : FastDerivation leq x y) ->
  Prerequisite (derivationType der) =>
  x `leq` y
CalcSmart (|~ x) = reflexive
CalcSmart (|~ x <~ y ... step) = step
CalcSmart ((der <~ z ... step1) {x,y}
               <~ w ... step2) @{prereq} =
                CalcSmart (der <~ z ... step1)
                @{inductivePrerequisite der (z ... step1) (w ... step2) prereq}
                  `transitive` step2

public export
CalcWith : Preorder dom leq => {0 x : dom} -> {0 y : dom} -> FastDerivation leq x y -> x `leq` y
CalcWith der = CalcSmart der @{preorderPrerequisite der}
```
This should also work with the existing code base, though I haven't tested it.
It would be good to test also if the `CalcSmart` constructor works instead of your (upcoming?) `TransitiveReasoning` module.
What do you think?

**Review by ohad — CHANGES_REQUESTED** [review]

**Comment by gallais** [comment]
Our testing should heavily involve incomplete proofs to ensure we do
not get hard errors when Idris cannot yet compute the `derivationType`
because it gets stuck on a hole.

**Comment by ohad** [comment]
I agree, though it might work out fine anyway:
+ The current interface (`CalcWith`) requires a `Preorder`, so that's not going to cause new errors.
+ The new interface (`CalcSmart`) computes the interface it needs based on the derivation structure, which ought to be known for incomplete proofs.

But I agree we ought to test these cases.

File: libs/base/Syntax/PreorderReasoning/Generic.idr [hunk 1]
```diff
@@ -18,10 +18,33 @@ data FastDerivation : (leq : a -> a -> Type) -> (x : a) -> (y : a) -> Type where
...
+Reflexive dom leq => CalcType dom leq TrivialDerivation where
+  CalcKnown (|~ _) {derivTypeEq = Refl} = reflexive
+  CalcKnown (_ <~ _) impossible
+
+public export
+Transitive dom leq => CalcType dom leq NonTrivialDerivation where
+  CalcKnown (|~ _) impossible
+  CalcKnown ((|~ _) <~ (_ ... step)) = step
+  CalcKnown (deriv@(_ <~ _) <~ (_ ... step)) = CalcKnown deriv `transitive` step
+
```
[the remark this answers is not in the document]
Comment (crisoagf) [reply libs/base/Syntax/PreorderReasoning/Generic.idr:43]: No, they are not necessary, I just wasn't aware of how to write `Prerequisite`!

Also `CalcSmart` is a way better name! I'll grab your changes!

**Comment by crisoagf** [comment]
> (...)
> 
> Do we really need to use type-classes to do this? A principle I learned from @gallais: don't search when you can compute. We know that we need a transitive relation for 2-steps or more, and we know we need a reflexive relation for 0-steps, and we know we don't need anything for 1 step: we already have the proof.
>
No, we don't, I just wasn't aware you could make constraints value-dependent! Grabbing your changes.
 
> 
> This should also work with the existing code base, though I haven't tested it. It would be good to test also if the `CalcSmart` constructor works instead of your (upcoming?) `TransitiveReasoning` module. What do you think?

With `CalcSmart` is there any need for `TransitiveReasoning`? I was thinking of dropping it given the new approach.

**Comment by crisoagf** [comment]
One other thing that could be done is just add the derivation type to the `FastDerivation` type constructor arguments. That would make transitive derivation and `~=` compatible (with this code if you use `~=` idris can't compute the derivation type, because it would require pattern matching on the `Refl`.

**Comment by crisoagf** [comment]
@ohad 

> Trawling the stdlib, [here](https://github.com/idris-lang/Idris2/blob/69f680e10a336c4f33414cfd55c13d41b68d735b/libs/contrib/Data/Nat/Division.idr#L271) is one instance where we're proving a strict inequality. It would be good to try and refactor it using the new CalcSmart mechanism and see if it works.

That proof isn't a great example to work on because it goes something like
```
n - 1
< n
== 1*n
<= (1+k) * n -- Note that this has to be a non-strict inequality, because k = 0 is possible
<= (1+k) * n + a -- Again, has to be non-strict, because a=0 is possible
```

Even replacing `n` with `predn + 1`, we still get the same issue.

I did not find any "good" chains of `transitive` to work with (other than some `Fin` proofs, but those don't work for other reasons), but I haven't stopped searching yet.

`minusLtMonotone` was promising, but it has 0 quantity on its arguments, which is not compatible with the `.Generic` module.

**Comment by ohad** [comment]
`LTE` is a proposition,  so you can implement 
```idris
recall : (0 prf : a `LTE` b) -> a `LTE` b
```
and lead with that

File: libs/base/Syntax/PreorderReasoning.idr [hunk 2]
```diff
@@ -52,3 +52,8 @@ public export
 public export
 (..>) : (0 y : a) -> (0 step : x ~=~ y) -> Step x y
 (..>) = (...)
+
+-- For documenting an equality that the compiler knows about but is not trivial to the reader.
```
Comment (ohad) [inline libs/base/Syntax/PreorderReasoning.idr:56]:
```suggestion
||| Use a judgemental equality but is not trivial to the reader.
```

**Review by ohad — APPROVED** [review]

**Comment by ohad** [comment]
Sorry this has taken so long. Thanks for implementing all the changes --- looks good! I'll merge once tests pass.

**Comment by ohad** [comment]
The issue with `test-elab` is that they're also using an infix operator `.=>`, but with different fixity precedence/association, and the importing of `PreorderReasoning` interferes with their conventions.

I opened issue #2832 that this kind of overriding ought to at least generate a warning, and @stefan-hoeck has agree to look into how to best change `elab-utils` to work around this issue.

**Comment by stefan-hoeck** [comment]
This has been fixed at elab-util, so the test should now succeed.

**Comment by ohad** [comment]
I'm told that to fix the broken test we should upstream changes from `main`:

https://github.com/crisoagf/Idris2/pull/1

**Comment by gallais** [comment]
It does not look like the elab-utils fix was successful?

**Comment by stefan-hoeck** [comment]
> It does not look like the elab-utils fix was successful?

I'll have a look.

**Comment by ohad** [comment]
Does [this PR helps](https://github.com/stefan-hoeck/idris2-elab-util/pull/64)?

**Comment by stefan-hoeck** [comment]
> Does [this PR helps](https://github.com/stefan-hoeck/idris2-elab-util/pull/64)?

Thanks for this, but I think a more thorough fix is OK. I deprecated the lambda operator an (to the best of my knowledge) removed all its uses from the code base. Please restart the CI tests. elab-util should be fine now.

**Comment by stefan-hoeck** [comment]
Damn. I'll look again.

**Comment by stefan-hoeck** [comment]
Sorry for this. Please run again. (Fingers crossed)

**Comment by ohad** [comment]
Awesome! Thanks everyone, and apologies to @crisoagf for the merge taking so long!

**Pull request merged**